File: leetcode/199.BinaryTreeRightSideView.go
package leetcode

func rightSideView(root *TreeNode) []int {
	var queue []*TreeNode
	var result []int

	queue = append(queue, root)

	for len(queue) > 0 {
		length := len(queue)
		var right *TreeNode

		for i := 0; i < length; i++ {
			node := queue[0]
			queue = queue[1:]

			if node != nil {
				right = node
				queue = append(queue, node.Left)
				queue = append(queue, node.Right)
			}
		}
		if right != nil {
			result = append(result, right.Val)
		}
	}

	return result
}

File: leetcode/199.BinaryTreeRightSideView.ts
import { TreeNode } from './shared'

function rightSideView(root: TreeNode | null): number[] {
  const result: number[] = []
  const queue: (TreeNode | null)[] = []

  queue.push(root)

  while (queue.length) {
    const length = queue.length
    let right = null
    for (let i = 0; i < length; i++) {
      const node = queue.shift()

      if (node) {
        right = node
        queue.push(node.left)
        queue.push(node.right)
      }
    }

    if (right) result.push(right.val)
  }

  return result
}
const tree = new TreeNode(1)
tree.left = new TreeNode(2)
tree.left.right = new TreeNode(5)
tree.right = new TreeNode(3)
tree.right.right = new TreeNode(4)

console.log(rightSideView(tree))
